pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ParseLock.java

```diff
@@ -116,4 +134,39 @@ private enum ParseStatus {
             this.isFinished = finished;
         }
     }
+
+    /**
+     * Subclasses of this assert that any thread can hold at most one lock at a time. Threads cannot even reenter in the
+     * lock they currently own. This prevents deadlocks while parsing ClassStub. However, all the instances of all derived
+     * subclasses are mutually exclusive. This is meant for the parse lock of ClassStub, and should therefore only be extended
+     * by that one.
+     */
+    abstract static class CheckedParseLock extends ParseLock {
+        private static final ThreadLocal<ParseLock> CURRENT_LOCK = new ThreadLocal<>();
+
+        protected CheckedParseLock(String name) {
+            super(name);
+        }
+
+        @Override
+        final void checkWeAreNotParsingAnother() {
+            if (isAssertEnabled()) {
```

Comment:
Why do we do this check only when asserts are enabled? Wouldn't it be better to always check, that we don't reenter/parse another class, while parsing? Then we would simply throw a IllegalStateException - as this is a bug.
Or is it so much slower using ThreadLocal, that his has a significant impact?

It seems, that we have `-ea` enabled for the regression-tester (https://github.com/pmd/pmd-regression-tester/blob/2684c6395d958d176eee659ebb776000c843b094/lib/pmdtester/builders/pmd_report_builder.rb#L96) - so we should at least see the errors (if any) there.

Reply:
I think it is just not possible that this branch is taken given how the code is written. It is just validating an assumption for our peace of mind. I don't think using the threadlocal significantly degrades performance but still to me this belongs to the kind of check that we will only want to do to debug a problem